### PR TITLE
[Serializer] avoid calling undefined built-in is_*() functions

### LINKED ISSUE #57314: [Serializer] Call to undefined function is_mixed() in AbstractObjectNormalizer.php line 780
### Symfony version(s) affected

7.1.1

### Description

#53160 introduced a bug in serializing objects with `mixed` type.

### How to reproduce

Serialize any object with at least one property with `mixed` type.

### Possible Solution

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by xabbuh**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Issues        | Fix #57314
| License       | MIT

### CODE REVIEW

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 1]
```diff
@@ -566,11 +566,23 @@ private function validateAndDenormalizeLegacy(array $types, string $currentClass
...
+                    LegacyType::BUILTIN_TYPE_FLOAT => \is_float($data),
+                    LegacyType::BUILTIN_TYPE_INT => \is_int($data),
+                    LegacyType::BUILTIN_TYPE_ITERABLE => is_iterable($data),
+                    LegacyType::BUILTIN_TYPE_NULL => null === $data,
+                    LegacyType::BUILTIN_TYPE_OBJECT => \is_object($data),
+                    LegacyType::BUILTIN_TYPE_RESOURCE => $data,
+                    LegacyType::BUILTIN_TYPE_STRING => \is_string($data),
+                    LegacyType::BUILTIN_TYPE_TRUE => true === $data,
+                    default => false,
+                };
```
Comment (alexandre-daubois) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: Would worth a private method maybe? As you're using the same logic twice?

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 2]
```diff
@@ -566,11 +566,23 @@ private function validateAndDenormalizeLegacy(array $types, string $currentClass
...
+                    LegacyType::BUILTIN_TYPE_ARRAY => \is_array($data),
+                    LegacyType::BUILTIN_TYPE_BOOL => \is_bool($data),
+                    LegacyType::BUILTIN_TYPE_CALLABLE => \is_callable($data),
+                    LegacyType::BUILTIN_TYPE_FALSE => false === $data,
+                    LegacyType::BUILTIN_TYPE_FLOAT => \is_float($data),
+                    LegacyType::BUILTIN_TYPE_INT => \is_int($data),
+                    LegacyType::BUILTIN_TYPE_ITERABLE => is_iterable($data),
+                    LegacyType::BUILTIN_TYPE_NULL => null === $data,
+                    LegacyType::BUILTIN_TYPE_OBJECT => \is_object($data),
+                    LegacyType::BUILTIN_TYPE_RESOURCE => $data,
```
Comment (derrabus) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]:
```suggestion
                    LegacyType::BUILTIN_TYPE_RESOURCE => \is_resource($data),
```

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php [hunk 3]
```diff
@@ -566,11 +566,23 @@ private function validateAndDenormalizeLegacy(array $types, string $currentClass
...
+                    LegacyType::BUILTIN_TYPE_STRING => \is_string($data),
+                    LegacyType::BUILTIN_TYPE_TRUE => true === $data,
+                    default => false,
```
Comment (derrabus) [inline src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: In the linked bug report, people explicitly mentioned the `mixed` type. I don't see it handled here and `false` as a fallback does not seem accurate: Anything can be mapped to mixed, so `$dataMatchesExpectedType` should be always true if we encounter `mixed`, shouldn't it?

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php (hunk 1 above)
Comment (xabbuh) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: In fact, the legacy type doesn't support mixed here. So I have reverted these changes.

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php (hunk 3 above)
Comment (xabbuh) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: Yes, you are right. I was able to extract a test case and adding what you suggested solves the issue.

**Review by derrabus — APPROVED** [review]

**Review by stof — APPROVED** [review]

File: src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php (hunk 1 above)
Comment (stof) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: @xabbuh the bug is not only for `mixed`. See https://github.com/symfony/symfony/issues/57334 reporting the issue for `false` (and probably also `true`). And the legacy type does support `false`

Comment (xabbuh) [reply src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php]: @stof You are right, but we need to fix this on the `5.4` branch (see #57341).

**Comment by HectorOrdonez** [comment]
Hello! When are we planning to release this? It's a breaking change for us.

**Comment by xabbuh** [comment]
Patch releases are published once a month.

**Pull request merged**